### PR TITLE
Integration of DPFIsolation and DeepTauId 

### PR DESCRIPTION
**Pull request by MRD2F**
First implementation of DPFIsolation and DeepTauId integrated into cmssw.

### CODE REVIEW

File: RecoTauTag/RecoTau/python/DPFIsolation_cff.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+import FWCore.ParameterSet.Config as cms
+
+from RecoTauTag.RecoTau.DPFIsolation_cff import *
```
Comment (mbluj) [inline RecoTauTag/RecoTau/python/DPFIsolation_cff.py]: Should be `DPFIsolation_cfi` rather than `cff`

**Review by mbluj — CHANGES_REQUESTED** [review]
Please provide a python configuration file in `RecoTauTag/RecoTau/test` to produce pat::Tau collection with the new Tau-Ids.  It can be called `runDeepTauIDsOnMiniAOD.py` or similar.

File: RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# produce pat::Tau collection with the new DNN Tau-Ids from miniAOD 12Apr2018_94X_mc2017
+
+import FWCore.ParameterSet.Config as cms
+from FWCore.ParameterSet.VarParsing import VarParsing
+
+options = VarParsing('analysis')
+options.parseArguments()
+
+process = cms.Process('runDeepTauIDsOnMiniAOD')
```
Comment (mbluj) [inline RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py]: Please use a shorter name of process, e.g. `TauID` or so.

File: RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+process.options = cms.untracked.PSet()
+process.options.wantSummary = cms.untracked.bool(False)
+process.options.allowUnscheduled = cms.untracked.bool(True)
+process.options.numberOfThreads = cms.untracked.uint32(8)
+process.options.numberOfStreams = cms.untracked.uint32(0)
+
+process.load('FWCore.MessageLogger.MessageLogger_cfi')
```
Comment (mbluj) [inline RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py]: ll16-17: I propose to move customization of `MessageLogger` at the end of the file.

File: RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+process.MessageLogger.cerr.FwkReport.reportEvery = 100
```
Comment (mbluj) [inline RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py]: I propose make frequency of report variable with number of events to be processed, e.g. like this
```
if process.maxEvents.input.value()>10:
     process.MessageLogger.cerr.FwkReport.reportEvery = process.maxEvents.input.value()//10
if process.maxEvents.input.value()>10000 or process.maxEvents.input.value()<0:
     process.MessageLogger.cerr.FwkReport.reportEvery = 1000
```

File: RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+# produce pat::Tau collection with the new DNN Tau-Ids from miniAOD 12Apr2018_94X_mc2017
+
+import FWCore.ParameterSet.Config as cms
+from FWCore.ParameterSet.VarParsing import VarParsing
+
+options = VarParsing('analysis')
```
Comment (mbluj) [inline RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py]: Is `VarParsing` intended to be used in any way in this example? If not, I would remove it - please comment.

**Review by mbluj — APPROVED** [review]

**Comment by mbluj** [comment]
I find the code fine, so I merge it - thanks for your work. 
In addition I have had a set of small comments to the tauId embedding tool, i.e. `runTauIdMVA.py` and the test configuration `runDeepTauIDsOnMiniAOD.py`, but I found that it is simpler that I implement them myself. A pull request will be sent soon.

**Comment by mbluj** [comment]
Hello again,
I have just merged to `cms-tau-pog:CMSSW_9_4_X_tau_pog_DNNTauID` updates mentioned above. 
I think you can start work on forward-porting DNN-related developments to 10_2_X and 10_4_X. I think that other release series are not important and can be ignored. I let you know names of branches to PR soon. 
Finally, when 10_4_X version is ready I will also ask for integration of training files.

**Comment by mbluj** [comment]
Branches for 10_4_X and 10_2_X forwardports are respectively CMSSW_10_4_X_tau_pog_DNNTauIDs and CMSSW_10_2_X_tau_pog_DNNTauIDs

**Comment by MRD2F** [comment]
> Branches for 10_4_X and 10_2_X forwardports are respectively CMSSW_10_4_X_tau_pog_DNNTauIDs and CMSSW_10_2_X_tau_pog_DNNTauIDs

Ok thanks for the information, I'll start to work on that

**Comment by kandrosov** [comment]
@mbluj Looking through the code, I realized that current implementations of both DeepTauId and DPFIso are not threadsafe. Should this be fixed?

**Comment by mbluj** [comment]
On Wed, Oct 24, 2018 at 5:44 PM Konstantin Androsov <
notifications@github.com> wrote:

> @mbluj <https://github.com/mbluj> Looking through the code, I realized
> that current implementations of both DeepTauId and DPFIso are not
> threadsafe. Should this be fixed?
>
I have not noticed this, and yes it should be fixed.

**Comment by veelken** [comment]
Hi, 
I am wondering if we want to have the DeepTauId and DPFIso also for a dR=0.3 isolation cone - or is it better to commission the algorithm for dR=0.5 first and add a version for dR=0.3 later ?

**Comment by kandrosov** [comment]
> Hi,
> I am wondering if we want to have the DeepTauId and DPFIso also for a dR=0.3 isolation cone - or is it better to commission the algorithm for dR=0.5 first and add a version for dR=0.3 later ?

Hello,
in my opinion, it is better to fully integrate the current implementations of the DNN discriminators first. Once it is done, it would be much easier to add dR=0.3 or new improved dR=0.5 versions.

**Comment by mbluj** [comment]
Hi Christian,
I think it makes sense to test both deepTau and DPF as they are also with ttH->tau,tau signal and ttbar background. As far as I understand, both were trained with "dense" events in mind even if both of them use dR=0.5 rather than 0.3. Anyway, those discriminants are still experimental and feedback from ttH people is welcome. It is planned that they are provided for wider tests soon (say next week).

**Comment by kandrosov** [comment]
> On Wed, Oct 24, 2018 at 5:44 PM Konstantin Androsov < ***@***.***> wrote: @mbluj <https://github.com/mbluj> Looking through the code, I realized that current implementations of both DeepTauId and DPFIso are not threadsafe. Should this be fixed?
> I have not noticed this, and yes it should be fixed.

Sorry, I have missed this problem as well... @MRD2F will create another PR that will address this issue.

**Comment by mbluj** [comment]
On Wed, Oct 24, 2018 at 6:17 PM Konstantin Androsov <
notifications@github.com> wrote:

> On Wed, Oct 24, 2018 at 5:44 PM Konstantin Androsov < ***@***.***> wrote:
> @mbluj <https://github.com/mbluj> https://github.com/mbluj Looking
> through the code, I realized that current implementations of both DeepTauId
> and DPFIso are not threadsafe. Should this be fixed?
> I have not noticed this, and yes it should be fixed.
>
> Sorry, I have missed this problem as well... @MRD2F
> <https://github.com/MRD2F> will create another PR that will address this
> issue.
>
OK, thanks.

**Pull request merged**